[Don't merge] Release 3.1.0 (#670)
apache/pulsar-site: release-notes/versioned/client-java-3.1.0.md

---
id: client-java-3.1.0
title: Client Java 3.1.0
sidebar_label: Client Java 3.1.0
---

- Disable polling pattern topics when TopicListWatcher is enabled. [#20779](https://github.com/apache/pulsar/pull/20779)

> - fix negative message re-delivery twice issue [#20750](https://github.com/apache/pulsar/pull/20750)

Reviewer: ```suggestion
- Fix negative message re-delivery twice issue [#20750](https://github.com/apache/pulsar/pull/20750)
```

- Fix subscribing pattern topics through Proxy not working [#20739](https://github.com/apache/pulsar/pull/20739)

> - Messages lost when consumer reconnect [#20695](https://github.com/apache/pulsar/pull/20695)

Reviewer: ```suggestion
- Messages lost when consumers reconnect [#20695](https://github.com/apache/pulsar/pull/20695)
```

> - Fix deadlock issue of consumer while using multiple IO threads [#20669](https://github.com/apache/pulsar/pull/20669)

Reviewer: ```suggestion
- Fix the deadlock issue of consumers while using multiple IO threads [#20669](https://github.com/apache/pulsar/pull/20669)
```

- Make the whole grabCnx() progress atomic [#20595](https://github.com/apache/pulsar/pull/20595)
- Swallow Conscrypt ClassNotFoundException [#20371](https://github.com/apache/pulsar/pull/20371)

> - thread-safe seek [#20321](https://github.com/apache/pulsar/pull/20321)

Reviewer: Can you re-phrase this issue?
